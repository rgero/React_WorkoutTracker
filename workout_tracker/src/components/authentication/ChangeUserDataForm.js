import React, {useState} from 'react';

import Alert from 'react-bootstrap/Alert';
import Button from 'react-bootstrap/Button';
import Container from 'react-bootstrap/Container';
import Col from 'react-bootstrap/Col';
import Form from 'react-bootstrap/Form';
import Row from 'react-bootstrap/Row';


export const ChangeUserFormData = ({currentUser, onSubmit})=> 
{
    const [email, setEmail] = useState(currentUser ? currentUser.email : "");
    const [displayName, setName] = useState(currentUser ? currentUser.displayName : "");
    const [currentPassword, setCurrentPassword] = useState("");
    const [newPassword, setNewPassword] = useState("");
    const [newPasswordRetype, setRetypePassword] = useState("");
    const [errMsg, setErrorMessage] = useState("")

    const trySignUp = async (event) => {
        event.preventDefault();
        if (currentPassword === "")
        {
            setErrorMessage("You need to fill out your current password");
        }

        if ( email === "" || displayName === "")
        {
            setErrorMessage("You must have an email and a display name.");
            return;
        }
        if (newPassword !== newPasswordRetype)
        {
            setErrorMessage("Passwords do not match.");
            return;
        }
        setErrorMessage("");

        let changes = {};
        if (newPassword !== "")
        {
            changes.password = newPassword;
        }
        changes.displayName = displayName;
        changes.email = email;

        await onSubmit(currentUser.email, currentPassword, changes);
    }

    return(
        <Container fluid="md">
            <Row className="justify-content-md-center">
                <Col>
                    <h3>User Profile</h3>
                    <Form onSubmit={trySignUp}>
                        <Form.Group className="mb-3" controlId="formEmail">
                            <Form.Label>E-mail</Form.Label>
                            <Form.Control
                                aria-label="email"
                                type="email"
                                value={email}
                                onChange={e => setEmail(e.target.value)}
                            />
                        </Form.Group>
                        <Form.Group className="mb-3" controlId="displayName">
                            <Form.Label>Display Name</Form.Label>
                            <Form.Control
                                aria-label="displayName"
                                type="text"
                                value={displayName}
                                onChange={e => setName(e.target.value)}
                            />
                        </Form.Group>
                        <Form.Group className="mb-3" controlId="formPassword">
                            <Form.Label>Current Password</Form.Label>
                            <Form.Control
                                aria-label="currentPassword"
                                type="password"
                                value={currentPassword}
                                onChange={e => setCurrentPassword(e.target.value)}
                            />
                        </Form.Group>
                        <Form.Group className="mb-3" controlId="formPassword">
                            <Form.Label>Enter new password</Form.Label>
                            <Form.Control
                                aria-label="reentry"
                                type="password"
                                value={newPassword}
                                onChange={e => setNewPassword(e.target.value)}
                            />
                        </Form.Group>
                        <Form.Group className="mb-3" controlId="formPassword">
                            <Form.Label>Re-enter new password</Form.Label>
                            <Form.Control
                                aria-label="reentry"
                                type="password"
                                value={newPasswordRetype}
                                onChange={e => setRetypePassword(e.target.value)}
                            />
                        </Form.Group>
                        <Button variant="outline-secondary" type="submit">
                            Submit changes
                        </Button>
                        { errMsg ? 
                            (
                                <Container className="pt-4">
                                    <Alert variant='danger'>Error: {errMsg}</Alert>
                                </Container>
                            ) : null
                        }
                    </Form>
                </Col>
            </Row>
        </Container>
    )
}