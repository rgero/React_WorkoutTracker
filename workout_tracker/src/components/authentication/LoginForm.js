import React, {useState} from 'react';

import Alert from 'react-bootstrap/Alert';
import Button from 'react-bootstrap/Button';
import Container from 'react-bootstrap/Container';
import Col from 'react-bootstrap/Col';
import Form from 'react-bootstrap/Form';
import Row from 'react-bootstrap/Row';


export const LoginForm = ({onSubmit})=> 
{
    const [email, setEmail] = useState("");
    const [password, setPassword] = useState("");
    const [errMsg, setErrorMessage] = useState("");

    const trySignIn = async (event) => {
        event.preventDefault();
        if ( email === "")
        {
            setErrorMessage("Error: Missing e-mail");
            return;
        }

        if ( password === "")
        {
            setErrorMessage("Error: Missing password");
            return;
        }
        setErrorMessage("");
        await onSubmit({email, password});
    }

    return(
        <Container fluid="md">
            <Row>
                <Col xs lg="3">
                    <Form onSubmit={trySignIn}>
                        <Form.Group className="mb-3" controlId="formEmail">
                            <Form.Label>E-mail</Form.Label>
                            <Form.Control
                                aria-label="email"
                                type="email"
                                value={email}
                                onChange={e => setEmail(e.target.value)}
                            />
                        </Form.Group>

                        <Form.Group className="mb-3" controlId="formPassword">
                            <Form.Label>Password</Form.Label>
                            <Form.Control
                                aria-label="password"
                                type="password"
                                value={password}
                                onChange={e => setPassword(e.target.value)}
                            />
                        </Form.Group>
                        <Button variant="outline-secondary" type="submit">
                            Log In
                        </Button>
                        { errMsg ? 
                            (
                                <Container className="pt-4">
                                    <Alert variant='danger'>{errMsg}</Alert>
                                </Container>
                            ) : null
                        }
                    </Form>
                </Col>
            </Row>
        </Container>
    )
}