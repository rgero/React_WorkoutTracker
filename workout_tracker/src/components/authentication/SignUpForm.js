import React, {useState} from 'react';

import Alert from 'react-bootstrap/Alert';
import Button from 'react-bootstrap/Button';
import Container from 'react-bootstrap/Container';
import Col from 'react-bootstrap/Col';
import Form from 'react-bootstrap/Form';
import Row from 'react-bootstrap/Row';


export const SignUpForm = ({onSubmit})=> 
{
    const [email, setEmail] = useState("");
    const [password, setPassword] = useState("");
    const [passwordSecond, setRetypePassword] = useState("");
    const [errMsg, setErrorMessage] = useState("")

    const trySignUp = async (event) => {
        event.preventDefault();
        if ( email === "" || password === "" || passwordSecond === "")
        {
            setErrorMessage("Please fill out the full form.");
            return;
        }
        if (passwordSecond !== password)
        {
            setErrorMessage("Passwords do not match.");
            return;
        }
        setErrorMessage("");
        await onSubmit({email, password});
    }

    return(
        <Container fluid="md">
            <Row className="justify-content-md-center">
                <Col>
                    <h3>Sign Up</h3>
                    <Form onSubmit={trySignUp}>
                        <Form.Group className="mb-3" controlId="formEmail">
                            <Form.Label>E-mail</Form.Label>
                            <Form.Control
                                aria-label="email"
                                type="email"
                                value={email}
                                onChange={e => setEmail(e.target.value)}
                            />
                        </Form.Group>
                        <Form.Group className="mb-3" controlId="formPassword">
                            <Form.Label>Password</Form.Label>
                            <Form.Control
                                aria-label="password"
                                type="password"
                                value={password}
                                onChange={e => setPassword(e.target.value)}
                            />
                        </Form.Group>
                        <Form.Group className="mb-3" controlId="formPassword">
                            <Form.Label>Re-enter password</Form.Label>
                            <Form.Control
                                aria-label="reentry"
                                type="password"
                                value={passwordSecond}
                                onChange={e => setRetypePassword(e.target.value)}
                            />
                        </Form.Group>
                        <Button variant="outline-secondary" type="submit">
                            Sign Up
                        </Button>
                        { errMsg ? 
                            (
                                <Container className="pt-4">
                                    <Alert variant='danger'>Error: {errMsg}</Alert>
                                </Container>
                            ) : null
                        }
                    </Form>
                </Col>
            </Row>
        </Container>
    )
}