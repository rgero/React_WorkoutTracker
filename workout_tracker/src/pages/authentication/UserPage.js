import React from "react";

import Container from 'react-bootstrap/Container';
import Col from 'react-bootstrap/Col';
import Row from 'react-bootstrap/Row';

import {Context as AuthContext} from '../../context/AuthContext';

export const UserPage = () => {
    const {state, tryLocalSignin} = React.useContext(AuthContext);

    React.useEffect(()=> {
        tryLocalSignin();
    }, [])

    return (
        <Container className="pt-3">
            <Row>
                <Col><h2>User Profile</h2></Col>
            </Row>
            <Row>
                <Col>
                    Display Name
                </Col>
                <Col>
                    {state.displayName}
                </Col>
            </Row>
            <Row>
                <Col>
                    E-Mail
                </Col>
                <Col>
                    {state.email}
                </Col>
            </Row>
        </Container>
    );
};

export default UserPage;