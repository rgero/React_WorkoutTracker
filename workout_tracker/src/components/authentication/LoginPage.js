import React, { useContext, useEffect, useState } from 'react';
import Container from 'react-bootstrap/Container';
import { Navigate } from "react-router-dom";
import {Context as AuthContext} from '../../context/AuthContext';

import { LoginForm } from './LoginForm';


export const LoginPage = ()=> 
{
    const {state, signIn, clearErrorMessage, tryLocalSignin} = useContext(AuthContext);

    useEffect(()=> {
        tryLocalSignin();
    }, [])

    if (state.token)
    {
        clearErrorMessage();
        return (<Navigate replace to="/dashboard" />);
    } else 
    {
        return (
            <Container fluid="md">
                <h1>Login</h1>
                <LoginForm
                    errorMessage={state.errorMessage}
                    onSubmit={({email, password})=> signIn({email, password})}
                />
            </Container>
        )
    }
}