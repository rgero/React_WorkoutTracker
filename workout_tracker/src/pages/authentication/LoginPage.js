import React, { useContext, useEffect } from 'react';
import { Navigate } from "react-router-dom";

import Alert from 'react-bootstrap/Alert';
import Container from 'react-bootstrap/Container';
import Col from 'react-bootstrap/Col';
import Row from 'react-bootstrap/Row';

import {Context as AuthContext} from '../../context/AuthContext';
import { LoginForm } from '../../components/authentication/LoginForm';

export const LoginPage = ()=> 
{
    const {state, signIn, clearErrorMessage, tryLocalSignin} = useContext(AuthContext);

    useEffect(()=> {
        const tryToSignIn = async () => {
            await tryLocalSignin();
            if (state.token)
            {
                clearErrorMessage();
            }
        }

        tryToSignIn();
    }, [])

    if (state.token)
    {
        return (<Navigate replace to="/dashboard" />);
    } else 
    {
        return (
            <Container>
                <Row>
                    <Col md={8} className="d-none d-lg-block">
                        <img src="/images/dumbbells.jpg" width="100%" />
                    </Col>
                    <Col xs={10} md={4}>
                        <LoginForm
                            onSubmit={({email, password})=> signIn({email, password})}
                        />
                        { state.errorMessage ? (
                                <Container className="pt-4">
                                    <Alert variant='danger'>Error: {state.errorMessage}</Alert>
                                </Container>
                            ) : null
                        }
                    </Col>
                </Row>
            </Container>
        )
    }
}

export default LoginPage;