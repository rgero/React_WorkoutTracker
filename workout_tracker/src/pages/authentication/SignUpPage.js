import React, { useContext } from 'react';
import { useNavigate } from 'react-router-dom';

import Container from 'react-bootstrap/Container';
import Col from 'react-bootstrap/Col';
import Row from 'react-bootstrap/Row';

import {Context as AuthContext} from '../../context/AuthContext';
import { SignUpForm } from '../../components/authentication/SignUpForm';

export const SignUpPage = ()=> 
{
    const navigate = useNavigate();
    const {state, signUp} = useContext(AuthContext);

    if (state.token)
    {
        navigate('/dashboard');
    } else 
    {
        return (
            <Container fluid="md">
                <Row>
                    <Col>
                        <SignUpForm onSubmit={({email, password})=> signUp({email, password})}
                    />
                    </Col>
                </Row>

            </Container>
        )
    }
}