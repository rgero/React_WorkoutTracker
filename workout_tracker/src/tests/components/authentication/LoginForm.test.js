import TestRenderer from 'react-test-renderer';

import {render, screen} from '@testing-library/react';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event'
import { act } from 'react-dom/test-utils';

import {LoginForm} from '../../../components/authentication/LoginForm';

describe("LoginForm Rendering Tests", ()=> {
    test("No Data", ()=> {
        let element = new TestRenderer.create(
            <LoginForm onSubmit={jest.fn()} />
        )
        expect(element.toJSON()).toMatchSnapshot();
    })
})

describe("LoginForm value tests", ()=> {
    let mockSubmit = jest.fn();
    let user;
    let mockUserData = {
        email: "gibgibgib@gib.com",
        password: "NotARealPasswordObvs"
    }

    beforeEach(()=> {
        user = userEvent.setup();
        render(<LoginForm onSubmit={mockSubmit} />)
    })

    test("Valid Submission", async ()=> {
        const emailInput = screen.getByRole("textbox", {name: /email/i});
        const passwordInput = screen.getByLabelText(/password/i);

        await user.click(emailInput);
        await user.keyboard(mockUserData.email);
    
        await user.click(passwordInput);
        await user.keyboard(mockUserData.password);
    
        const button = screen.getByRole("button");
        await user.click(button);

        expect(mockSubmit).toHaveBeenCalled();
        expect(mockSubmit).toHaveBeenCalledWith( mockUserData )
    })

    test("Invalid Submission, no email", async ()=> {
        const passwordInput = screen.getByLabelText(/password/i);

        await user.click(passwordInput);
        await user.keyboard(mockUserData.password);
        
        const button = screen.getByRole("button");
        await user.click(button);

        expect(mockSubmit).not.toHaveBeenCalled();

        const errorField = screen.getByText("Error: Missing e-mail");
        expect(errorField).toBeInTheDocument();
    })

    test("Invalid Submission - missing password", async ()=> {
        const emailInput = screen.getByRole("textbox", {name: /email/i});

        await user.click(emailInput);
        await user.keyboard(mockUserData.email);
                
        const button = screen.getByRole("button");
        await user.click(button);

        expect(mockSubmit).not.toHaveBeenCalled();

        const errorField = screen.getByText("Error: Missing password");
        expect(errorField).toBeInTheDocument();
    })
})