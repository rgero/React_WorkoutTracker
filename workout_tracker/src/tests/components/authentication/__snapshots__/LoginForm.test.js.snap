// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm Rendering Tests No Data 1`] = `
<div
  className="container-md"
>
  <div
    className="justify-content-md-center row"
  >
    <div
      className="col"
    >
      <h3>
        Log in
      </h3>
      <form
        className=""
        onSubmit={[Function]}
      >
        <div
          className="mb-3"
        >
          <label
            className="form-label"
            htmlFor="formEmail"
          >
            E-mail
          </label>
          <input
            aria-label="email"
            className="form-control"
            id="formEmail"
            onChange={[Function]}
            type="email"
            value=""
          />
        </div>
        <div
          className="mb-3"
        >
          <label
            className="form-label"
            htmlFor="formPassword"
          >
            Password
          </label>
          <input
            aria-label="password"
            className="form-control"
            id="formPassword"
            onChange={[Function]}
            type="password"
            value=""
          />
        </div>
        <button
          className="btn btn-outline-secondary"
          disabled={false}
          type="submit"
        >
          Log In
        </button>
      </form>
    </div>
  </div>
</div>
`;
